### PR TITLE
JENKINS-60999: Add support for a new parameter tailMaxCount for BUILD_LOG_REGEX token

### PR DESCRIPTION
**Pull request by haridsv**
Included a fix (and test cases) for a edge case scenario of using `greedy=false` with default `maxCount`. Also added missing documentation for `greedy`.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java [hunk 1]
```diff
@@ -234,15 +240,17 @@ String getContent(BufferedReader reader)
...
         }
-        insidePre = stopPre(buffer, insidePre);
-        if (buffer.length() == 0) {
+        insidePre = stopPre(matchResults, insidePre);
+        if (matchResults.size() == 0) {
             return defaultValue;
         }
-        return buffer.toString();
+        if (maxTailMatches > 0 && matchResults.size() > maxTailMatches) {
+            matchResults = matchResults.subList(matchResults.size() - maxTailMatches, matchResults.size());
```
Comment (slide) [inline src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java]: What happens if it is in the middle of a `<pre></pre>` and cuts off? See the insidePre checks above. I think we would want to make sure that a `</pre>` gets added at the end if one is needed.

Comment (haridsv) [reply src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java]: Thanks for pointing that out, let me look into the right handling for this case.

File: src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java [hunk 2]
```diff
@@ -234,15 +244,72 @@ String getContent(BufferedReader reader)
...
-        insidePre = stopPre(buffer, insidePre);
-        if (buffer.length() == 0) {
+        preStartLoc = stopPre(matchResults, preStartLoc, preRanges);
+        if (matchResults.size() == 0) {
             return defaultValue;
         }
-        return buffer.toString();
+        if (maxTailMatches > 0 && matchResults.size() > maxTailMatches) {
+            int tailStartLocation;
+            if (asHtml) {
```
Comment (haridsv) [inline src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java]: @slide I was considering to handle this like how the truncated lines are handled in the else case, but this will become unreliable if there are pre blocks in the log output itself, so decided to not do it. Please let me know if you think otherwise.

File: src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java (hunk 1 above)
Comment (haridsv) [reply src/main/java/org/jenkinsci/plugins/tokenmacro/impl/BuildLogRegexMacro.java]: I addressed your concern, please let me know if you have any other feedback.
